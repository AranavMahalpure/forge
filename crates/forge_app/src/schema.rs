// @generated automatically by Diesel CLI.

diesel::table! {
    configuration_table (id) {
        id -> Text,
        data -> Text,
        created_at -> Timestamp,
    }
}

diesel::table! {
    conversations (id) {
        id -> Text,
        created_at -> Timestamp,
        updated_at -> Timestamp,
        content -> Text,
        archived -> Bool,
        title -> Nullable<Text>,
    }
}

diesel::table! {
    learning_table (id) {
        id -> Text,
        cwd -> Text,
        learnings -> Text,
        created_at -> Timestamp,
    }
}

diesel::allow_tables_to_appear_in_same_query!(
    configuration_table,
    conversations,
    learning_table,
);
