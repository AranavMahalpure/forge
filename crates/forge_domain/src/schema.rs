/// Generate a JSON schema for a type without the `$schema` field.
#[macro_export]
macro_rules! json_schema {
    ($t:ty) => {{
        let mut settings = ::schemars::gen::SchemaSettings::default();
        settings.meta_schema = None;
        let gen = ::schemars::gen::SchemaGenerator::new(settings);
        gen.into_root_schema_for::<$t>()
    }};
}


#[cfg(test)]
mod tests {
    use schemars::JsonSchema;

    #[test]
    fn test(){
        #[derive(JsonSchema)]
        struct Test {
            _name: String,
        }
        insta::assert_snapshot!(serde_json::to_string_pretty(&json_schema!(Test)).unwrap());
    }
}