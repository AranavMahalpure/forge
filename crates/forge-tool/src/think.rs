use std::collections::HashMap;

use crate::{
    model::{CallToolRequest, CallToolResponse, ToolResponseContent, ToolsListResponse},
    Tool,
};
use anyhow::Result;
use colorize::AnsiColor;
use serde::{Deserialize, Serialize};
use serde_json::json;

#[derive(Debug, Serialize, Deserialize, Clone)]
struct ThoughtData {
    thought: String,
    thought_number: i32,
    total_thoughts: i32,
    next_thought_needed: bool,
    is_revision: Option<bool>,
    revises_thought: Option<i32>,
    branch_from_thought: Option<i32>,
    branch_id: Option<String>,
    needs_more_thoughts: Option<bool>,
}

#[derive(Clone)]
struct SequentialThinkingServer {
    thought_history: Vec<ThoughtData>,
    branches: HashMap<String, Vec<ThoughtData>>,
}

impl SequentialThinkingServer {
    fn new() -> Self {
        Self {
            thought_history: Vec::new(),
            branches: HashMap::new(),
        }
    }

    fn validate_thought_data(&self, input: serde_json::Value) -> Result<ThoughtData> {
        let thought_data: ThoughtData = serde_json::from_value(input)?;

        if thought_data.thought_number <= 0 {
            return Err(anyhow::anyhow!("Invalid thoughtNumber: must be positive"));
        }
        if thought_data.total_thoughts <= 0 {
            return Err(anyhow::anyhow!("Invalid totalThoughts: must be positive"));
        }

        Ok(thought_data)
    }

    fn format_thought(&self, thought_data: &ThoughtData) -> String {
        let (prefix, context) = match (thought_data.is_revision, &thought_data.branch_from_thought)
        {
            (Some(true), _) => (
                "🔄 Revision".yellow().to_string(),
                format!(
                    " (revising thought {})",
                    thought_data.revises_thought.unwrap_or(0)
                ),
            ),
            (_, Some(branch)) => (
                "🌿 Branch".green().to_string(),
                format!(
                    " (from thought {}, ID: {})",
                    branch,
                    thought_data.branch_id.as_ref().unwrap_or(&String::new())
                ),
            ),
            _ => ("💭 Thought".blue().to_string(), String::new()),
        };

        let header = format!(
            "{} {}/{}{}",
            prefix, thought_data.thought_number, thought_data.total_thoughts, context
        );
        let border_len = header.len().max(thought_data.thought.len()) + 4;
        let border = "─".repeat(border_len);

        format!(
            "\n┌{}┐\n│ {} │\n├{}┤\n│ {} │\n└{}┘",
            border,
            header,
            border,
            format!("{:width$}", thought_data.thought, width = border_len - 2),
            border
        )
    }

    fn process_thought(&mut self, input: serde_json::Value) -> Result<serde_json::Value> {
        let mut thought_data = self.validate_thought_data(input)?;

        if thought_data.thought_number > thought_data.total_thoughts {
            thought_data.total_thoughts = thought_data.thought_number;
        }

        self.thought_history.push(thought_data.clone());

        if let (Some(branch_from), Some(branch_id)) =
            (thought_data.branch_from_thought, &thought_data.branch_id)
        {
            self.branches
                .entry(branch_id.clone())
                .or_default()
                .push(thought_data.clone());
        }

        eprintln!("{}", self.format_thought(&thought_data));

        let result = serde_json::json!({
            "thoughtNumber": thought_data.thought_number,
            "totalThoughts": thought_data.total_thoughts,
            "nextThoughtNeeded": thought_data.next_thought_needed,
            "branches": self.branches.keys().collect::<Vec<_>>(),
            "thoughtHistoryLength": self.thought_history.len()
        });

        Ok(result)
    }
}

fn call_tool(
    req: CallToolRequest,
    thinking_server: &mut SequentialThinkingServer,
) -> Result<CallToolResponse> {
    let name = req.name.as_str();
    let args = req.arguments.unwrap_or_default();
    let args = serde_json::to_value(args)?;
    let result = match name {
        "sequentialthinking" => {
            let thought_result = thinking_server.process_thought(args)?;
            ToolResponseContent::Text {
                text: serde_json::to_string(&thought_result)?,
            }
        }
        _ => return Err(anyhow::anyhow!("Unknown tool: {}", req.name)),
    };
    Ok(CallToolResponse {
        content: vec![result],
        is_error: None,
        meta: None,
    })
}

fn list_tools() -> Result<ToolsListResponse> {
    let response = json!({
      "tools": [
        {
          "name": "sequentialthinking",
          "description": "A detailed tool for dynamic and reflective problem-solving through thoughts.",
          "inputSchema": {
            "type": "object",
            "properties": {
              "thought": {
                "type": "string",
                "description": "Your current thinking step"
              },
              "nextThoughtNeeded": {
                "type": "boolean",
                "description": "Whether another thought step is needed"
              },
              "thoughtNumber": {
                "type": "integer",
                "description": "Current thought number",
                "minimum": 1
              },
              "totalThoughts": {
                "type": "integer",
                "description": "Estimated total thoughts needed",
                "minimum": 1
              },
              "isRevision": {
                "type": "boolean",
                "description": "Whether this revises previous thinking"
              },
              "revisesThought": {
                "type": "integer",
                "description": "Which thought is being reconsidered",
                "minimum": 1
              },
              "branchFromThought": {
                "type": "integer",
                "description": "Branching point thought number",
                "minimum": 1
              },
              "branchId": {
                "type": "string",
                "description": "Branch identifier"
              },
              "needsMoreThoughts": {
                "type": "boolean",
                "description": "If more thoughts are needed"
              }
            },
            "required": ["thought", "nextThoughtNeeded", "thoughtNumber", "totalThoughts"]
          },
        }
      ],
    });
    Ok(serde_json::from_value(response)?)
}

struct Think {
    thinking_server: SequentialThinkingServer,
}

impl Default for Think {
    fn default() -> Self {
        Self {
            thinking_server: SequentialThinkingServer::new(),
        }
    }
}

#[async_trait::async_trait]
impl Tool for Think {
    fn name(&self) -> &'static str {
        "Sequential Thinking"
    }

    async fn tools_call(&self, input: CallToolRequest) -> Result<CallToolResponse, String> {
        call_tool(input, &mut self.thinking_server.clone()).map_err(|e| e.to_string())
    }

    fn tools_list(&self) -> Result<ToolsListResponse, String> {
        list_tools().map_err(|e| e.to_string())
    }
}
