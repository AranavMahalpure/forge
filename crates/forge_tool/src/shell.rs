use std::collections::HashSet;
use std::path::PathBuf;

use anyhow::Result;
use forge_domain::{Description, ToolCallService};
use forge_tool_macros::Description;
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use tokio::process::Command;

#[derive(Debug, Serialize, Deserialize, Clone, JsonSchema)]
pub struct ShellInput {
    /// The shell command to execute.
    pub command: String,
    /// The working directory where the command should be executed.
    pub cwd: PathBuf,
}

#[derive(Debug, Serialize, Deserialize, Clone, JsonSchema)]
pub struct ShellOutput {
    pub stdout: String,
    pub stderr: String,
    pub success: bool,
}

/// Execute shell commands with safety checks and validation. This tool provides
/// controlled access to system shell commands while preventing dangerous
/// operations through a comprehensive blacklist and validation system.
#[derive(Description)]
pub struct Shell {
    blacklist: HashSet<String>,
}

impl Default for Shell {
    fn default() -> Self {
        let mut blacklist = HashSet::new();
        // File System Destruction Commands
        blacklist.insert("rm".to_string());
        blacklist.insert("rmdir".to_string());
        blacklist.insert("del".to_string());

        // Disk/Filesystem Commands
        blacklist.insert("format".to_string());
        blacklist.insert("mkfs".to_string());
        blacklist.insert("dd".to_string());

        // Permission/Ownership Commands
        blacklist.insert("chmod".to_string());
        blacklist.insert("chown".to_string());

        // Privilege Escalation Commands
        blacklist.insert("sudo".to_string());
        blacklist.insert("su".to_string());

        // Code Execution Commands
        blacklist.insert("exec".to_string());
        blacklist.insert("eval".to_string());

        // System Communication Commands
        blacklist.insert("write".to_string());
        blacklist.insert("wall".to_string());

        // System Control Commands
        blacklist.insert("shutdown".to_string());
        blacklist.insert("reboot".to_string());
        blacklist.insert("init".to_string());

        Shell { blacklist }
    }
}

impl Shell {
    fn validate_command(&self, command: &str) -> Result<(), String> {
        let base_command = command
            .split_whitespace()
            .next()
            .ok_or_else(|| "Empty command".to_string())?;

        if self.blacklist.contains(base_command) {
            return Err(format!("Command '{}' is not allowed", base_command));
        }

        Ok(())
    }

    async fn execute_command(&self, command: &str, cwd: PathBuf) -> Result<ShellOutput, String> {
        let mut cmd = if cfg!(target_os = "windows") {
            let mut c = Command::new("cmd");
            c.args(["/C", command]);
            c
        } else {
            let mut c = Command::new("sh");
            c.args(["-c", command]);
            c
        };

        cmd.current_dir(cwd);

        let output = cmd.output().await.map_err(|e| e.to_string())?;

        Ok(ShellOutput {
            stdout: String::from_utf8_lossy(&output.stdout).to_string(),
            stderr: String::from_utf8_lossy(&output.stderr).to_string(),
            success: output.status.success(),
        })
    }
}

#[async_trait::async_trait]
impl ToolCallService for Shell {
    type Input = ShellInput;
    type Output = ShellOutput;

    async fn call(&self, input: Self::Input) -> Result<Self::Output, String> {
        self.validate_command(&input.command)?;
        self.execute_command(&input.command, input.cwd).await
    }
}

#[cfg(test)]
mod tests {
    use std::{env, fs};

    use pretty_assertions::assert_eq;

    use super::*;

    #[tokio::test]
    async fn test_shell_echo() {
        let shell = Shell::default();
        let result = shell
            .call(ShellInput {
                command: "echo 'Hello, World!'".to_string(),
                cwd: env::current_dir().unwrap(),
            })
            .await
            .unwrap();

        assert!(result.success);
        assert!(result.stdout.contains("Hello, World!"));
        assert!(result.stderr.is_empty());
    }

    #[tokio::test]
    async fn test_shell_with_working_directory() {
        let shell = Shell::default();
        let temp_dir = fs::canonicalize(env::temp_dir()).unwrap();

        let result = shell
            .call(ShellInput {
                command: if cfg!(target_os = "windows") {
                    "cd".to_string()
                } else {
                    "pwd".to_string()
                },
                cwd: temp_dir.clone(),
            })
            .await
            .unwrap();

        assert!(result.success);
        let output_path = PathBuf::from(result.stdout.trim());
        let actual_path = match fs::canonicalize(output_path.clone()) {
            Ok(path) => path,
            Err(_) => output_path,
        };
        let expected_path = temp_dir.as_path();

        assert_eq!(
            actual_path, expected_path,
            "\nExpected path: {:?}\nActual path: {:?}",
            expected_path, actual_path
        );
        assert!(result.stderr.is_empty());
    }

    #[tokio::test]
    async fn test_shell_invalid_command() {
        let shell = Shell::default();
        let result = shell
            .call(ShellInput {
                command: "nonexistentcommand".to_string(),
                cwd: env::current_dir().unwrap(),
            })
            .await
            .unwrap();

        assert!(!result.success);
        assert!(!result.stderr.is_empty());
    }

    #[tokio::test]
    async fn test_shell_blacklisted_command() {
        let shell = Shell::default();
        let result = shell
            .call(ShellInput {
                command: "rm -rf /".to_string(),
                cwd: env::current_dir().unwrap(),
            })
            .await;

        assert!(result.is_err());
        assert!(result.unwrap_err().contains("not allowed"));
    }

    #[tokio::test]
    async fn test_shell_empty_command() {
        let shell = Shell::default();
        let result = shell
            .call(ShellInput { command: "".to_string(), cwd: env::current_dir().unwrap() })
            .await;

        assert!(result.is_err());
        assert!(result.unwrap_err().contains("Empty command"));
    }

    #[tokio::test]
    async fn test_shell_pwd() {
        let shell = Shell::default();
        let current_dir = fs::canonicalize(env::current_dir().unwrap()).unwrap();
        let result = shell
            .call(ShellInput {
                command: if cfg!(target_os = "windows") {
                    "cd".to_string()
                } else {
                    "pwd".to_string()
                },
                cwd: current_dir.clone(),
            })
            .await
            .unwrap();

        assert!(result.success);
        let output_path = PathBuf::from(result.stdout.trim());
        let actual_path = match fs::canonicalize(output_path.clone()) {
            Ok(path) => path,
            Err(_) => output_path,
        };
        assert_eq!(actual_path, current_dir);
        assert!(result.stderr.is_empty());
    }

    #[tokio::test]
    async fn test_shell_multiple_commands() {
        let shell = Shell::default();
        let result = shell
            .call(ShellInput {
                command: "echo 'first' && echo 'second'".to_string(),
                cwd: env::current_dir().unwrap(),
            })
            .await
            .unwrap();

        assert!(result.success);
        assert!(result.stdout.contains("first"));
        assert!(result.stdout.contains("second"));
        assert!(result.stderr.is_empty());
    }

    #[tokio::test]
    async fn test_shell_with_environment_variables() {
        let shell = Shell::default();
        let result = shell
            .call(ShellInput {
                command: "echo $PATH".to_string(),
                cwd: env::current_dir().unwrap(),
            })
            .await
            .unwrap();

        assert!(result.success);
        assert!(!result.stdout.is_empty());
        assert!(result.stderr.is_empty());
    }

    #[test]
    fn test_description() {
        assert!(Shell::description().len() > 100)
    }
}
