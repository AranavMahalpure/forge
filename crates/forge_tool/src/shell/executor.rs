use std::io::{self, Write};
use std::path::PathBuf;

use tokio::io::AsyncRead;
use tokio::process::Command;

/// A command executor that handles command creation and execution
#[derive(Debug)]
pub struct CommandExecutor {
    command: Command,
    colored: bool,
}

pub struct Output {
    pub stdout: String,
    pub stderr: String,
    pub success: bool,
}

impl CommandExecutor {
    /// Create a new command executor with the specified command and working
    /// directory
    pub fn new(command: &str, cwd: &PathBuf) -> Self {
        let mut command = if cfg!(target_os = "windows") {
            let mut c = Command::new("cmd");
            c.args(["/C", command]);
            c
        } else {
            let mut c = Command::new("sh");
            c.args(["-c", command]);
            c
        };
        // Set the current working directory for the command
        command.current_dir(cwd);
        // Kill the command when the handler is dropped
        command.kill_on_drop(true);
        Self { command, colored: false }
    }

    /// Enable colored output for the command
    pub fn colored(mut self) -> Self {
        self.command.env("CLICOLOR_FORCE", "1");
        self.colored = true;
        self
    }

    /// executes the command and streams the output of command to stdout and
    /// stderr.
    pub async fn execute(mut self) -> anyhow::Result<Output> {
        // don't change this configuration else it will break the cli output.
        self.command
            .stdin(std::process::Stdio::inherit())
            .stdout(std::process::Stdio::piped())
            .stderr(std::process::Stdio::piped());

        let mut child = self.command.spawn()?;

        let mut stdout_pipe = child.stdout.take();
        let mut stderr_pipe = child.stderr.take();

        let stdout_fut = stream(&mut stdout_pipe, io::stdout());
        let stderr_fut = stream(&mut stderr_pipe, io::stderr());

        let (status, stdout, stderr) = tokio::try_join!(child.wait(), stdout_fut, stderr_fut)?;
        // Drop happens after `try_join` due to <https://github.com/tokio-rs/tokio/issues/4309>
        drop(stdout_pipe);
        drop(stderr_pipe);

        let (stdout, stderr) = if self.colored {
            (
                String::from_utf8(strip_ansi_escapes::strip(stdout))?,
                String::from_utf8(strip_ansi_escapes::strip(stderr))?,
            )
        } else {
            (String::from_utf8(stdout)?, String::from_utf8(stderr)?)
        };

        Ok(Output { success: status.success(), stdout, stderr })
    }
}

// streams the output from child process to stdout and stderr.
async fn stream<A: AsyncRead + Unpin, W: Write>(
    io: &mut Option<A>,
    mut writer: W,
) -> io::Result<Vec<u8>> {
    let mut output = Vec::new();
    use tokio::io::AsyncReadExt;
    if let Some(io) = io.as_mut() {
        let mut buff = [0; 1024];
        loop {
            let n = io.read(&mut buff).await?;
            if n == 0 {
                break;
            }
            writer.write_all(&buff[..n])?;
            writer.flush()?;
            output.extend_from_slice(&buff[..n]);
        }
    }
    Ok(output)
}
