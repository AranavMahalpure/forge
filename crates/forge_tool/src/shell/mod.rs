mod executor;
mod shell;
mod streamer;

pub use shell::*;
