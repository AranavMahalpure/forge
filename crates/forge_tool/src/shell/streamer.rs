use std::io::{self, Write};

use tokio::io::AsyncReadExt;
use tokio::process::Child;

/// An output stream handler that captures and processes command output
pub struct OutputStream {
    buffer: Vec<u8>,
    writer: Box<dyn Write + Send>,
}

impl OutputStream {
    fn new(writer: Box<dyn Write + Send>) -> Self {
        Self { buffer: Vec::new(), writer }
    }

    /// Process stream data from a reader
    async fn process<T: tokio::io::AsyncRead + Unpin>(
        &mut self,
        mut reader: T,
    ) -> Result<(), String> {
        let mut buf = [0u8; 1024];

        loop {
            match reader.read(&mut buf).await {
                Ok(0) => break, // EOF
                Ok(n) => {
                    self.writer
                        .write_all(&buf[..n])
                        .map_err(|e| format!("Failed to write to stream: {}", e))?;
                    self.buffer.extend_from_slice(&buf[..n]);
                }
                Err(e) => return Err(format!("Failed to read from stream: {}", e)),
            }
        }

        Ok(())
    }

    /// Convert the captured output to a string, removing ANSI escape codes
    fn into_string(self) -> Result<String, String> {
        String::from_utf8(strip_ansi_escapes::strip(self.buffer))
            .map_err(|e| format!("Failed to convert output to string: {}", e))
    }
}

/// A stream processor that handles command output streams
pub struct CommandStreamer {
    child: Child,
    stdout: OutputStream,
    stderr: OutputStream,
}

impl CommandStreamer {
    pub fn new(child: Child) -> Self {
        // Initialize output streams without taking ownership yet
        let stdout = OutputStream::new(Box::new(io::stdout()));
        let stderr = OutputStream::new(Box::new(io::stderr()));

        Self { child, stdout, stderr }
    }

    /// Stream and process command output
    pub async fn stream(mut self) -> Result<(String, String, bool), String> {
        // Get stream handles
        let stdout = self
            .child
            .stdout
            .take()
            .ok_or_else(|| "Child process stdout not configured".to_string())?;
        let stderr = self
            .child
            .stderr
            .take()
            .ok_or_else(|| "Child process stderr not configured".to_string())?;

        // Process streams concurrently
        tokio::try_join!(self.stdout.process(stdout), self.stderr.process(stderr))?;

        // Wait for command completion
        let status = self
            .child
            .wait()
            .await
            .map_err(|e| format!("Failed to wait for command: {}", e))?;

        Ok((
            self.stdout.into_string()?,
            self.stderr.into_string()?,
            status.success(),
        ))
    }
}
