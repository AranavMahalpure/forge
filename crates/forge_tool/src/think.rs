use std::collections::HashMap;

use anyhow::Result;
use colorize::AnsiColor;
use forge_domain::{NamedTool, ToolCallService, ToolDescription, ToolName};
use forge_tool_macros::ToolDescription;
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

/// Framework for tracking and managing problem-solving steps. Supports
/// step-by-step task breakdown, revisions, branching, and confidence tracking.
/// Use `total_thoughts` for complexity, `revises_thought` for revisions,
/// `branch_from_thought` for branching, and `solution_confidence` for progress.
/// Track steps in `thought_history`, manage alternate paths in `branches`, mark
/// completion with `solution_reached
#[derive(Clone, Default, ToolDescription)]
pub struct Think {
    thought_history: Vec<ThoughtInput>,
    branches: HashMap<String, Vec<ThoughtInput>>,
    solution_reached: bool,
}

#[derive(Debug, Serialize, Deserialize, Clone, JsonSchema)]
pub struct ThoughtInput {
    /// Description of current thought/reasoning step
    pub thought: String,
    /// Indicates if another thought is needed
    pub next_thought_needed: bool,
    /// Number of current thought/step
    pub thought_number: i32,
    /// Total thoughts/steps expected for solution
    pub total_thoughts: i32,
    /// Indicates if this is a revision
    #[serde(skip_serializing_if = "Option::is_none")]
    pub is_revision: Option<bool>,
    /// Number of thought being revised
    #[serde(skip_serializing_if = "Option::is_none")]
    pub revises_thought: Option<i32>,
    /// Parent thought number for this branch
    #[serde(skip_serializing_if = "Option::is_none")]
    pub branch_from_thought: Option<i32>,
    /// Unique branch identifier
    #[serde(skip_serializing_if = "Option::is_none")]
    pub branch_id: Option<String>,
    /// Indicates if more thoughts needed
    #[serde(skip_serializing_if = "Option::is_none")]
    pub needs_more_thoughts: Option<bool>,
    /// Solution confidence (0.0-1.0)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub solution_confidence: Option<f32>,
}

#[derive(Debug, Serialize, Deserialize, Clone, JsonSchema)]
pub struct ThoughtResult {
    pub thought_number: i32,
    pub total_thoughts: i32,
    pub next_thought_needed: bool,
    pub solution_reached: bool,
    pub solution_confidence: f32,
    pub branches: Vec<String>,
    pub thought_history_length: usize,
}

impl Think {
    fn validate_thought_data(&self, mut input: ThoughtInput) -> Result<ThoughtInput> {
        if input.thought_number <= 0 {
            return Err(anyhow::anyhow!("Invalid thoughtNumber: must be positive"));
        }
        if input.total_thoughts <= 0 {
            return Err(anyhow::anyhow!("Invalid totalThoughts: must be positive"));
        }

        // If no confidence is provided, calculate it based on progress
        if input.solution_confidence.is_none() {
            input.solution_confidence =
                Some(input.thought_number as f32 / input.total_thoughts as f32);
        }

        Ok(input)
    }

    fn format_thought(&self, thought_data: &ThoughtInput) -> String {
        let (prefix, context) = match (thought_data.is_revision, &thought_data.branch_from_thought)
        {
            (Some(true), _) => (
                "🔄 Revision".yellow().to_string(),
                format!(
                    " (revising thought {})",
                    thought_data.revises_thought.unwrap_or(0)
                ),
            ),
            (_, Some(branch)) => (
                "🌿 Branch".green().to_string(),
                format!(
                    " (from thought {}, ID: {})",
                    branch,
                    thought_data.branch_id.as_ref().unwrap_or(&String::new())
                ),
            ),
            _ => ("💭 Thought".blue().to_string(), String::new()),
        };

        let header = format!(
            "{} {}/{}{} (Confidence: {:.2}%)",
            prefix,
            thought_data.thought_number,
            thought_data.total_thoughts,
            context,
            thought_data.solution_confidence.unwrap_or(0.0) * 100.0
        );
        let border_len = header.len().max(thought_data.thought.len()) + 4;
        let border = "─".repeat(border_len);

        let thought_data = format!("{:width$}", thought_data.thought, width = border_len - 2);

        format!(
            "\n┌{}┐\n│ {} │\n├{}┤\n│ {} │\n└{}┘",
            border, header, border, thought_data, border
        )
    }

    fn process_thought(&mut self, input: ThoughtInput) -> Result<ThoughtResult> {
        let mut thought_data = self.validate_thought_data(input)?;

        // Adjust total thoughts if needed
        if thought_data.thought_number > thought_data.total_thoughts {
            thought_data.total_thoughts = thought_data.thought_number;
        }

        // Evaluate solution confidence
        if let Some(confidence) = thought_data.solution_confidence {
            if confidence >= 0.8 {
                self.solution_reached = true;
                thought_data.next_thought_needed = false;
            }
        }

        // Terminate thinking if max thoughts reached or solution found
        if thought_data.thought_number >= thought_data.total_thoughts || self.solution_reached {
            thought_data.next_thought_needed = false;
        }

        // Always allow at least one thought to be processed
        if self.thought_history.is_empty() {
            thought_data.next_thought_needed = true;
        }

        self.thought_history.push(thought_data.clone());

        // Branch handling remains the same
        if let (Some(_), Some(branch_id)) =
            (thought_data.branch_from_thought, &thought_data.branch_id)
        {
            self.branches
                .entry(branch_id.clone())
                .or_default()
                .push(thought_data.clone());
        }

        eprintln!("{}", self.format_thought(&thought_data));

        Ok(ThoughtResult {
            thought_number: thought_data.thought_number,
            total_thoughts: thought_data.total_thoughts,
            next_thought_needed: thought_data.next_thought_needed,
            solution_reached: self.solution_reached,
            solution_confidence: thought_data.solution_confidence.unwrap_or(0.0),
            branches: self.branches.keys().cloned().collect(),
            thought_history_length: self.thought_history.len(),
        })
    }
}

impl NamedTool for Think {
    fn tool_name(&self) -> ToolName {
        ToolName::new("think_step")
    }
}

#[async_trait::async_trait]
impl ToolCallService for Think {
    type Input = ThoughtInput;
    type Output = ThoughtResult;

    async fn call(&self, input: Self::Input) -> Result<Self::Output, String> {
        let mut thinker = self.clone();
        let thought_result = thinker.process_thought(input).map_err(|e| e.to_string())?;
        Ok(thought_result)
    }
}
