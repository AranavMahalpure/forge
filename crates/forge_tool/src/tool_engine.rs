use std::collections::HashMap;

use inflector::Inflector;
use schemars::schema::RootSchema;
use schemars::{schema_for, JsonSchema};
use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::shell::Shell;
use crate::think::Think;
use crate::{
    AskFollowUpQuestion, Description, FSFileInfo, FSList, FSRead, FSReplace, FSSearch, FSWrite,
    Outline, ToolTrait,
};

struct JsonTool<T>(T);

#[async_trait::async_trait]
impl<T: ToolTrait + Sync> ToolTrait for JsonTool<T>
where
    T::Input: serde::de::DeserializeOwned + JsonSchema,
    T::Output: serde::Serialize + JsonSchema,
{
    type Input = Value;
    type Output = Value;

    async fn call(&self, input: Self::Input) -> Result<Self::Output, String> {
        let input: T::Input = serde_json::from_value(input).map_err(|e| e.to_string())?;
        let output: T::Output = self.0.call(input).await?;
        Ok(serde_json::to_value(output).map_err(|e| e.to_string())?)
    }
}

struct ToolDefinition {
    executable: Box<dyn ToolTrait<Input = Value, Output = Value> + Send + Sync + 'static>,
    tool: Tool,
}

pub struct ToolEngine {
    tools: HashMap<ToolName, ToolDefinition>,
}

///
/// Refer to the specification over here:
/// https://glama.ai/blog/2024-11-25-model-context-protocol-quickstart#server
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Tool {
    pub name: ToolName,
    pub description: String,
    pub input_schema: RootSchema,
    pub output_schema: Option<RootSchema>,
}

#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(transparent)]
pub struct ToolName(pub String);

impl ToolName {
    pub fn into_string(self) -> String {
        self.0
    }

    pub fn as_str(&self) -> &str {
        &self.0
    }
}

impl ToolEngine {
    pub async fn call(&self, tool_id: &ToolName, input: Value) -> Result<Value, String> {
        match self.tools.get(tool_id) {
            Some(tool) => tool.executable.call(input).await,
            None => Err(format!("No such tool found: {}", tool_id.as_str())),
        }
    }

    pub fn list(&self) -> Vec<Tool> {
        self.tools.values().map(|tool| tool.tool.clone()).collect()
    }

    fn import<T>(tool: T) -> (ToolName, ToolDefinition)
    where
        T: ToolTrait + Description + Send + Sync + 'static,
        T::Input: serde::de::DeserializeOwned + JsonSchema,
        T::Output: serde::Serialize + JsonSchema,
    {
        let id = std::any::type_name::<T>()
            .split("::")
            .last()
            .unwrap()
            .to_snake_case();
        let executable = Box::new(JsonTool(tool));
        let tool = Tool {
            name: ToolName(id.clone()),
            description: T::description().to_string(),
            input_schema: schema_for!(T::Input),
            output_schema: Some(schema_for!(T::Output)),
        };

        (ToolName(id), ToolDefinition { executable, tool })
    }
}

impl Default for ToolEngine {
    fn default() -> Self {
        let tools: HashMap<ToolName, ToolDefinition> = HashMap::from([
            ToolEngine::import(FSRead),
            ToolEngine::import(FSSearch),
            ToolEngine::import(FSList),
            ToolEngine::import(FSFileInfo),
            ToolEngine::import(FSWrite),
            ToolEngine::import(FSReplace),
            ToolEngine::import(Outline),
            ToolEngine::import(Think::default()),
            ToolEngine::import(Shell::default()),
            ToolEngine::import(AskFollowUpQuestion),
        ]);

        Self { tools }
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_id() {
        assert!(ToolEngine::import(FSRead)
            .0
            .into_string()
            .ends_with("fs_read"));
        assert!(ToolEngine::import(FSSearch)
            .0
            .into_string()
            .ends_with("fs_search"));
        assert!(ToolEngine::import(FSList)
            .0
            .into_string()
            .ends_with("fs_list"));
        assert!(ToolEngine::import(FSFileInfo)
            .0
            .into_string()
            .ends_with("file_info"));
        assert!(ToolEngine::import(Think::default())
            .0
            .into_string()
            .ends_with("think"));
    }
}
