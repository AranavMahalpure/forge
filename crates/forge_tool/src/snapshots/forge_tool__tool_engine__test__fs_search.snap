---
source: crates/forge_tool/src/tool_engine.rs
expression: tool_str
---
{
  "name": "fs_search",
  "description": "Request to perform a regex search across files in a specified directory, providing context-rich results. This tool searches for patterns or specific content across multiple files, displaying each match with encapsulating context. Parameters: - path: (required) The path of the directory to search in (relative to the current working directory ./test-dir). This directory will be recursively searched. - regex: (required) The regular expression pattern to search for. Uses Rust regex syntax. - file_pattern: (optional) Glob pattern to filter files (e.g., \\'*.ts\\' for TypeScript files). If not provided, it will search all files (*). Usage: <search_files> <path>Directory path here</path> <regex>Your regex pattern here</regex> <file_pattern>file pattern here (optional)</file_pattern> </search_files>",
  "input_schema": {
    "$schema": "http://json-schema.org/draft-07/schema#",
    "title": "FSSearchInput",
    "type": "object",
    "required": [
      "path",
      "regex"
    ],
    "properties": {
      "file_pattern": {
        "description": "Glob pattern to filter files (e.g., '*.ts' for TypeScript files). If not provided, it will search all files (*).",
        "type": [
          "string",
          "null"
        ]
      },
      "path": {
        "description": "The path of the directory to search in (relative to the current working directory ./test-dir). This directory will be recursively searched.",
        "type": "string"
      },
      "regex": {
        "description": "The regular expression pattern to search for. Uses Rust regex syntax.",
        "type": "string"
      }
    }
  },
  "output_schema": {
    "$schema": "http://json-schema.org/draft-07/schema#",
    "title": "Array_of_String",
    "type": "array",
    "items": {
      "type": "string"
    }
  }
}
